src/deny-imports.ts

```diff
@@ -0,0 +1,99 @@
+import path from "node:path"
+
+import { normalizePath, type PluginOption, type ResolvedConfig } from "vite"
+
+import { type Env } from "./env"
+import { type Pattern, findMatch } from "./pattern"
+
+type EnvPatterns = Partial<Record<Env, Pattern[]>>
+
+type Options = Record<Env, { specifiers?: Pattern[]; files?: Pattern[] }>
+
+export default function (options: Options): PluginOption[] {
+  let denySpecifiers: EnvPatterns = {}
+  let denyFiles: EnvPatterns = {}
+
+  for (let [env, { specifiers, files }] of Object.entries(options)) {
+    denySpecifiers[env as Env] = specifiers
+    denyFiles[env as Env] = files
+  }
+
+  return [denyImportSpecifiers(denySpecifiers), denyImportFiles(denyFiles)]
+}
+
+function denyImportSpecifiers(denySpecifiers: EnvPatterns): PluginOption {
+  let root: string
+  let command: ResolvedConfig["command"]
+  return {
+    name: "vite-plugin-deny-imports-specifiers",
+    enforce: "pre",
+    configResolved(config) {
+      root = config.root
+      command = config.command
+    },
+    resolveId(id, importer, options) {
+      if (!importer) return
+
+      let env: Env = options?.ssr ? "server" : "client"
+      let denialPattern = findMatch(id, denySpecifiers[env])
+      if (denialPattern) {
+        let message = [
+          `Import denied in ${env} environment`,
+          ` - Import: "${id}"`,
+          ` - Importer: ${normalizeRelativePath(root, importer)}`,
+          ` - Pattern: ${denialPattern}`,
+        ].join("\n")
+        throw Error(message)
```

Comment:
inlining this felt like a nice reduction in complexity, but it does mean we need to test the error messages with e2e tests that run Vite. But actually I think that's better since it more closely resembles real usage.